src/ui/AttachResult/AttachResult.ts

```diff
@@ -0,0 +1,180 @@
+            defaultFunction: () => () => new Promise<void>((resolve, reject) => { resolve(); })
+        })
+    };
+
+    constructor(
+        public element: HTMLElement,
+        public options: IAttachResultOptions,
+        public bindings?: IComponentBindings,
+        public queryResult?: IQueryResult
+    ) {
+        super(element, AttachResult.ID, bindings);
+
+        this.options = ComponentOptions.initComponentOptions(element, AttachResult, options);
+        this.queryResult = this.queryResult || this.resolveResult();
+
+        this.initialize();
+
+        this.bind.on(this.element, "click", this.toggleAttached);
+    }
+
+    /**
+     * Gets whether or not the result is currently attached.
+     */
+    public isAttached() : boolean {
+        return this.attached;
+    }
+
+    /**
+     * Attach the query result.
+     */
+    public attach() : Promise<void> {
+        if (this.attached || this.loading) {
+            return Promise.resolve();
+        }
+
+        this.setLoading(true);
+        return this.options.attach(this.queryResult)
+            .then(() => { this.attached = true; })
+            .then(() => { 
+                let customData: IAnalyticsCaseAttachMeta = {
+                    resultUriHash: this.queryResult.raw.urihash,
+                    author: this.queryResult.raw.author,
+                    articleID: null,
+                    caseID: null
+                };
+          
+                this.usageAnalytics.logCustomEvent<IAnalyticsCaseAttachMeta>(analyticsActionCauseList.caseAttach, customData, this.root);
+            })
+            .finally(() => { this.setLoading(false); });
+    }
+
+    /**
+     * Detach the query result.
+     */
+    public detach() : Promise<void> {
+        if (!this.attached && !this.loading) {
+            return Promise.resolve();
+        }
+
+        this.setLoading(true);
+        return this.options.detach(this.queryResult)
+            .then(() => { this.attached = false; })
+            .then(() => {
+                let customData: IAnalyticsCaseAttachMeta = {
+                    resultUriHash: this.queryResult.raw.urihash,
+                    author: this.queryResult.raw.author,
+                    articleID: null,
+                    caseID: null
+                };
+          
+                this.usageAnalytics.logCustomEvent<IAnalyticsCaseDetachMeta>(analyticsActionCauseList.caseDetach, customData, this.root);
+            })
+            .finally(() => { this.setLoading(false); });
+    }
+
+    /** Toggle the state of the component. If the current result is not attached, attach it, if not, detach it. */
+    public toggleAttached() : void {
+        if (this.attached) {
+            this.detach();
+        } else {
+            this.attach();
+        }
+    }
+
+    protected initialize() : void {
+        // Adds HTML elements needed to display this component.
```

Comment:
I would remove this comment. It is better to reserve them for "why" you are doing something rather than "what" you are doing.